### ISSUE TITLE
issues in saving processed nifti file

### ISSUE THREAD

**Issue by xuesongle** [muschellij2/oro.nifti#4]
Hi, John:

I am testing with brain extraction algorithm in nifit format. The current issue is in saving the processed data. Would you be able to provide some assistance here? Thanks.

````
ibrary(oro.nifti)
setwd("~/R_demo")
inPath= "data/Template-T1-U8-RALPFH-BR.nii.gz"
img <- readNIfTI(inPath, reorient = FALSE)
class(img)
dim(img)
image(img, z = 225, plot.type = "single")
orthographic(img)
library(fslr)
fsldir = '/home/griffith/fsl'
options(fsl.path = fsldir)
get.fsl()

img_bet <- fslbet(infile = img, retimg = TRUE)
mask <- img
in_mask <- img_bet > 0
mask[in_mask] <- 1
mask[!in_mask] <- NA
orthographic(x = img, y = mask, col.y = "lightblue")

library(scales)
ortho2(img, mask, col.y = alpha("red", 0.3))


brain <- img
brain[!in_mask] <- NA
double_ortho(img, brain)

writeNIfTI(brain, "random")
list.files(pattern="random")
````
The error message occurred at "writeNIfTI(brain, "random")"
````
> writeNIfTI(brain, "random")
Error in .writeNIfTI(nim, filename, onefile, gzipped, verbose, warn, compression) : 
  all dim elements > dim[1] must be 1
> list.files(pattern="random")
character(0)
````
The dimension of the brain is:
> dim(brain)
[1] 182 512 512
>

**Comment by muschellij2** [comment]
If you have a question, please provide a MCVE: https://stackoverflow.com/help/mcve.  Can you send the image?

Please see `dim_` and `pixdim` functoins.

**Comment by xuesongle** [comment]
Thanks, @muschellij2 
The image is from [MRI Brain Template](http://www.jannin.org/mritemplate/data/Template-T1-U8-RALPFH-BR.nii.gz), 
[Template-T1-U8-RALPFH-BR.nii.gz](https://github.com/muschellij2/oro.nifti/files/2436189/Template-T1-U8-RALPFH-BR.nii.gz)

**Comment by muschellij2** [comment]
TL; DR - use `neurobase::readnii` and `neurobase::writenii` instead.

## Why is this happening
It is reasons like this that I recommend using `neurobase::readnii` and `neurobase::writenii`, which will fix these common problems.  These will not likely be fixed in `oro.nifti`, as these other functions exist and things like this can identify header problems after reading in with `readNIfTI`.

```r
> dim_(img)
```
```
[1]   3 182 512 512   0   0   0   0
```


``` r
setwd("~/Downloads/")

library(fslr)
#> Loading required package: oro.nifti
#> oro.nifti 0.9.8
#> Loading required package: neurobase
library(neurobase)
inPath= "Template-T1-U8-RALPFH-BR.nii.gz"
img <- readNIfTI(inPath, reorient = FALSE)
dim_(img)
#> [1]   3 182 512 512   0   0   0   0
i2 = readnii(inPath)
dim_(i2)
#> [1]   3 182 512 512   1   1   1   1

img_bet <- fslbet(infile = img, retimg = TRUE)
#> Warning in get.fsl(): Setting fsl.path to /usr/local/fsl
#> Warning in get.fsloutput(): Can't find FSLOUTPUTTYPE, setting to NIFTI_GZ
#> FSLDIR='/usr/local/fsl'; PATH=${FSLDIR}/bin:${PATH};export PATH FSLDIR; sh "${FSLDIR}/etc/fslconf/fsl.sh"; FSLOUTPUTTYPE=NIFTI_GZ; export FSLOUTPUTTYPE; ${FSLDIR}/bin/bet2 "/private/var/folders/1s/wrtqcpxn685_zk570bnx9_rr0000gr/T/RtmptCeSms/file171735147833.nii.gz" "/var/folders/1s/wrtqcpxn685_zk570bnx9_rr0000gr/T//RtmptCeSms/file17172d18c49e"
mask <- img
in_mask <- img_bet > 0
mask[in_mask] <- 1
mask[!in_mask] <- NA

brain <- img
brain[!in_mask] <- NA
dim_(brain)
#> [1]   3 182 512 512   0   0   0   0

brain = drop_img_dim(brain)
dim_(brain)
#> [1]   3 182 512 512   1   1   1   1
tfile = tempfile()
writeNIfTI(brain, tfile)
file.exists(paste0(tfile, ".nii.gz"))
#> [1] TRUE
```

Created on 2018-10-01 by the [reprex
package](http://reprex.tidyverse.org) (v0.2.0).

**Comment by xuesongle** [comment]
@muschellij2 . Many thanks for your prompt help and the solution. But please do not close it yet as I am reading each line of the code and definitely will bring more questions shortly.

**Comment by xuesongle** [comment]
@muschellij2 , a few simple questions but I still ask here. (Thanks for your patience and kindly help)
Question 1: I notice that readNIfTI does not exist in neurobase package. Does neurobase indirectly load oro.nifti package and call readNIfTI function. If that is the case, why not load ibrary(oro.nifti) instead? (I did run the search() and find both "package:neurobase" and "package:oro.nifti" attached to the path but not sure when)

Question 2: 
I notice that dim_(img)
#> [1]   3 182 512 512   0   0   0   0
I also noticed that i2 = readnii(inPath)
dim_(i2)
#> [1]   3 182 512 512   1   1   1   1
What does those number mean? I checked the documentation on Methods that act on the dim_ field in the NIfTI/ANALYZE header but no detailed explanation.

**Comment by muschellij2** [comment]
`readnii` is in `neurobase`.  `readnii` calls `readNIfTI`, but isn't the same behavior.  The `neurobase` package will load `oro.nifti` as well.

dim_ gives back the number of dimensions of an image `3`, then the dimensions , up to 7, usually `x` (182), `y` (512), `z` (512), `time`, then others.

**Comment by xuesongle** [comment]
@muschellij2 , does "0 0 0 0 " mean data is not loading properly while "1 1 1 1'  does?

**Comment by muschellij2** [comment]
It means that's what's in the header.  When writing, it looks to make sure
those are set to >= 1 (this is a misspecified header but not really
*wrong*).  If not, there is an error.  `readnii` fixes that with the
`drop_img_dim` function.
John

**Comment by xuesongle** [comment]
@muschellij2 , thanks John. so the drop_im_dim function set values in all dimensions 182 512 512 0 0 0 0 to 
#> [1]   3 182 512 512   1   1   1   1
Is my understanding correct?

**Comment by muschellij2** [comment]
It does other things, but correct.
John

**Comment by xuesongle** [comment]
@muschellij2 , all clear on this issue. Thanks again for your patience and help.
